data-structures/data.md

```diff
@@ -0,0 +1,30 @@
+`Flexible Byte Layout` is an advanced layout for representing binary data.
+
+It is flexible enough to support very small and very large (multi-block) binary data.
+
+```sh
+type Lengths [Int]
+type ByteUnionList [&FlexibleByteLayout]
+
+type NestedByteList struct {
+  lengths Lengths
+  parts ByteUnionList
+  algo optional String
```

Comment:
As I’m implementing this I’m wondering what the full implications of this are.

First of all, if two algorithms produce the same graph, this will cause them not to match.

Second, this isn’t really enough. For instance, I just implemented a balanced layout algorithm but it also accepts a limit (max number of links per block) and that is necessary to reproduce the algorithm. But if we add it, either as another property or as an addition to the string, it’ll make a lot of matching graphs (any two limits that are both above the total number in the FBL) no longer match.

🤷‍♂️

Reply:
Personally, I'd store the algorithm outside. E.g., when adding a file, I'd record how the file was imported in the metadata.

Reply:
this isn’t for the chunking algorithm, it’s for the layout algorithm used for balancing the tree.

Reply:
I'd store both in the file metadata.

Reply:
ya, let’s cut it from this spec and put it in a container. efficient mutations, which is the whole point of storing this info, could be considered out of scope here. this is all a lot simpler if the only role of an FBL is to represent an arbitrarily large series of order bytes.